genoml/preprocessing/munging.py

```diff
@@ -15,340 +15,369 @@
             subprocess.run(bash_rm_temp, shell=True)
-
-    # Checking the impute flag and execute
-        # Currently only supports mean and median
-        impute_list = ["mean", "median"]
-        
-        if (self.geno_path != None):
-
-            if impute_type not in impute_list:
-                return "The 2 types of imputation currently supported are 'mean' and 'median'"
-            elif impute_type.lower() == "mean":
-                raw_df = raw_df.fillna(raw_df.mean())
-            elif impute_type.lower() == "median":
-                raw_df = raw_df.fillna(raw_df.median())
-            print("")
-            print(
-                f"You have just imputed your genotype features, covering up NAs with the column {impute_type} so that analyses don't crash due to missing data.")
-            print("Now your genotype features might look a little better (showing the first few lines of the left-most and right-most columns)...")
-            print("#"*70)
-            print(raw_df.describe())
-            print("#"*70)
-            print("")
-
-    # Checking the imputation of non-genotype features
-
-        if (self.addit_path != None):
-            if impute_type not in impute_list:
-                return "The 2 types of imputation currently supported are 'mean' and 'median'"
-            elif impute_type.lower() == "mean":
-                addit_df = addit_df.fillna(addit_df.mean())
-            elif impute_type.lower() == "median":
-                addit_df = addit_df.fillna(addit_df.median())
-            print("")
+            # Checking the impute flag and execute
+            # Currently only supports mean and median
+            self.merged = _fill_impute_na(self.impute_type, self.merged)
+
+        # Checking the imputation of non-genotype features
+        if self.addit_path:
+            addit_df = self.munge_additional_features()
+
+        self.merged = _merge_dfs([self.pheno_df, self.merged, addit_df], col_id="ID")
+        del addit_df
+
+        self.merged = self.harmonize_refs(self.merged)
+        self.merged.to_hdf(self.output_datafile, key="dataForML")
+
+        features_list = self.merged.columns.tolist()
+
+        features_listpath = f"{self.run_prefix}.list_features.txt"
+        with open(features_listpath, "w") as f:
+            f.write("\n".join([str(feature) for feature in features_list]))
+
+        print(
+            f"An updated list of {len(features_list)} features, including ID and PHENO,"
+            f" that is in your munged dataForML.h5 file can be found here "
+            f"{features_listpath}"
+            "\n"
+            f"Your .dataForML file that has been fully munged can be found here: "
+            f"{self.output_datafile}"
+        )
+
+        return self.merged
+
+    def _run_external_plink_commands(self) -> None:
+        """Runs the external plink commands from the command line."""
+        if not self.geno_path:
+            return
+        cmds_a, cmds_b = get_bash_scripts(
```

Comment:
Should gwas_path be an argument here, so it can return only one set of cmds rather than both?  Seems like its being chosen between in the next if statement.

Reply:
I separated these two out to make it easier to check what the commands will be. Getting a list of strings in a single function logically is one action. Whereas if there is a gwas path, the dataframe is also reduced and output and different commands are used.

If they were to be one function, outside of the class, I would also have to pass in the `self.p_gwas` and `self.gwas_df`. But I can make them one function if you think that will read better.

Reply:
I see, so its intentionally comparative, yeah that's fine.  Maybe rename the local variables to match the follow if statement then, vs (a/b) that doesn't convey much information.